web/src/features/hfiCalculator/slices/hfiCalculatorSlice.ts

```diff
@@ -3,52 +3,278 @@ import { createSlice, PayloadAction } from '@reduxjs/toolkit'
+}
+
+export const calculatePrepLevel = (
+  meanIntensityGroup: number | undefined
+): number | undefined => {
+  // for now, prep level calculation assumed a fixed Fire Starts value of 0-1
+
+  if (isUndefined(meanIntensityGroup)) {
+    return undefined
+  } else {
+    meanIntensityGroup = Math.round(meanIntensityGroup)
+  }
+  if (meanIntensityGroup < 3) {
+    return 1
+  }
+  if (meanIntensityGroup < 4) {
+    return 2
+  }
+  if (meanIntensityGroup < 5) {
+    return 3
+  }
+  return 4
+}
+
+export const calculateMeanPrepLevel = (
+  rawMeanIntensityGroups: (number | undefined)[]
+): number | undefined => {
+  // for now, prep level calculation assumed a fixed Fire Starts value of 0-1
+  if (isUndefined(rawMeanIntensityGroups)) {
+    return undefined
+  } else {
+    const existingDailies: number[] = []
+    rawMeanIntensityGroups.forEach(daily => {
+      if (!isUndefined(daily)) {
+        existingDailies.push(Math.round(daily))
+      }
+    })
+    return Math.round(
+      existingDailies?.reduce((a, b) => a + b, 0) / existingDailies.length
+    )
+  }
+}
+
+// TODO: Inefficient, improve if it becomes a problem
+const calculateHFIResults = (
+  fireCentres: { [key: string]: FireCentre },
+  dailies: StationDaily[],
+  numPrepDays: number,
+  selected: number[],
+  selectedPrepDayIso: string
+): { [key: string]: HFIResult } => {
+  const selectedPrepDay =
+    selectedPrepDayIso == '' ? null : DateTime.fromISO(selectedPrepDayIso)
+  const planningAreaToDailies: { [key: string]: HFIResult } = {}
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  Object.entries(fireCentres).forEach(([_, fireCentre]) =>
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    Object.entries(fireCentre.planning_areas).forEach(([__, area]) => {
+      const areaStationCodes = new Set(
+        Object.entries(area.stations).map(([, station]) => station.code)
+      )
+      const areaDailies = dailies.filter(
+        daily => selected.includes(daily.code) && areaStationCodes.has(daily.code)
+      )
+
+      const dailyMeanIntensity = calculateMeanIntensity(
+        isNull(selectedPrepDayIso)
+          ? areaDailies
+          : areaDailies.filter(
+              day =>
+                day.date.year === selectedPrepDay?.year &&
+                day.date.month === selectedPrepDay?.month &&
+                day.date.day === selectedPrepDay?.day
+            )
+      )
+      const dailyMeanIntensityGroups = calculateDailyMeanIntensities(
+        areaDailies,
+        numPrepDays
+      )
+      const maxMeanIntensityGroup = calculateMaxMeanIntensityGroup(
+        dailyMeanIntensityGroups
+      )
+      const dailyPrepLevel = calculatePrepLevel(dailyMeanIntensity)
+      const dailyPrepLevels = calculateDailyPrepLevels(dailyMeanIntensityGroups)
+      const meanPrepLevel = calculateMeanPrepLevel(dailyPrepLevels)
+      planningAreaToDailies[area.name] = {
+        dailies: areaDailies,
+        dailyMeanIntensity,
```

Comment:
wouldn't it be more type-safe to cast all these into a `HFIResult` object, as the declared type is in L143?

Reply:
The Typescript compiler infers the type in the assignment here, it's not necessary to specify, and will explode if the wrong keys / types are passed: https://www.typescriptlang.org/docs/handbook/type-inference.html